core/src/test/scala/unit/kafka/admin/DescribeConsumerGroupTest.scala

```
@@ -315,24 +271,40 @@ class DescribeConsumerGroupTest extends KafkaServerTestHarness {
     val consumerGroupCommand = new KafkaConsumerGroupService(opts)
 
     TestUtils.waitUntilTrue(() => {
-        try {
           val (state, assignments) = consumerGroupCommand.describeGroup()
           state == Some("Stable") &&
           assignments.isDefined &&
           assignments.get.count(_.group == group) == 2 &&
           assignments.get.count{ x => x.group == group && x.partition.isDefined} == 2 &&
           assignments.get.count{ x => x.group == group && !x.partition.isDefined} == 0
-        } catch {
-          case _: GroupCoordinatorNotAvailableException | _: IllegalArgumentException =>
-            // Do nothing while the group initializes
-            false
-          case e: Throwable =>
-            throw e
-        }
       }, "Expected two rows (one row per consumer) in describe group results.")
 
     consumerGroupCommand.close()
   }
+
+  @Test
+  def testDescribeGroupWithNewConsumerWithShortInitializationTimeout() {
+    // run one consumer in the group consuming from a single-partition topic
+    val executor = new ConsumerGroupExecutor(brokerList, 1, group, topic)
+
+    // set the group initialization timeout too low for the group to stabilize
+    val cgcArgs = Array("--bootstrap-server", brokerList, "--describe", "--group", "group", "--timeout", "10")
+    val opts = new ConsumerGroupCommandOptions(cgcArgs)
+    val consumerGroupCommand = new KafkaConsumerGroupService(opts)
+
+    try {
+      val (state, assignments) = consumerGroupCommand.describeGroup()
+      fail("The consumer group command should fail due to low initialization timeout")
+    } catch {
+      case e @ (_: GroupCoordinatorNotAvailableException | _: IllegalArgumentException) =>
```

Why is `IllegalArgumentException` acceptable?

---

There are two types of exceptions that can occur with a short timeout while the group initializes (like the one used in this unit test). One is `GroupCoordinatorNotAvailableException` for coordinator related errors ([here](https://github.com/apache/kafka/blob/913c09e4a94447a953cb55ccd702d7131d1764a8/core/src/main/scala/kafka/admin/AdminClient.scala#L72)). The other, `IllegalArgumentException`, occurs if the new group's protocol is not yet set ([here](https://github.com/apache/kafka/blob/913c09e4a94447a953cb55ccd702d7131d1764a8/core/src/main/scala/kafka/admin/AdminClient.scala#L177)). These two exceptions are the ones we are trying to circumvent here by retrying the corresponding API calls within the preset timeout. I hope it makes sense.

---

Hmm.. Not sure I follow. For the `IllegalArgumentException`, we would have to have a group which is not dead, not empty, and which has a protocol type not matching "consumer." Is that really possible in this test case? 

I'm also wondering if we should raise `TimeoutException` with `GroupCoordinatorNotAvailableException` as the cause for the first case?

---

Regarding the first point you raised, with a short 10 ms timeout used in the unit test `IllegalArgumentException` does not occur in my dev environment. But when I raise the timeout to 400 ms it raises that exception. Not sure what the best way to capture both in unit tests are because the proper timeout values (at least for `IllegalArgumentException`) for the unit test would probably be different from one environment to another. Any suggestion?

On your second point, do you mean when we reach the timeout we just throw a `TimeoutException` instead of continuing with the normal process (without any retries)? I guess it makes sense for user to see the impact or importance of the timeout value they can set. I'll try to update the PR with this. Please advise if I misunderstood your point.

Thanks.

---

In the case that the `IllegalArgumentException` is thrown, which of the checks is failing?

---

The protocol type is empty (`''`). The reported error is this:
> java.lang.IllegalArgumentException: Consumer Group group with protocol type '' is not a valid consumer group

---

That's kind of weird. What is the state of the group in that case?

---

State has an empty value too.
> State: '', protocol type: ''

With a timeout of 400 ms in the unit test I had to run it a few times before hitting the exception. Is this behavior unexpected?

---

@vahidhashemian Actually it occurs to me that the problem is probably that we are checking for a valid consumer group before checking the error code in the response. The check for a valid consumer group probably only makes sense if the error code is NONE.

---

You are right. Then perhaps we can bring an extra check into group validation to avoid this error?
```
metadata.error == Errors.NONE && (metadata.state == "Dead" || metadata.state == "Empty" || metadata.protocolType == ConsumerProtocol.PROTOCOL_TYPE)
```

After a quick check, with the addition of a timeout exception we won't be needing to throw an `IllegalArgumentException` exception anymore. I'll submit an update shortly.